### PR TITLE
kvm: fix direct download template size

### LINKED ISSUE #8037: Unable to deploy a vm from a direct download templates on 4.19 release
ISSUE TYPE

* Bug

**COMPONENT NAME**

Component: templates

CLOUDSTACK VERSION

Cloudstack version 4.19

**SUMMARY**

Unable to deploy a vm from a direct download templates

**Steps to reproduce the issue**

1. Register a template mark it as Direct download

Screenshot 2023-10-04 at 1 54 52 PM

2. Launch a vm with the Direct download template

3. The VM deployment fails with the exception 
```

2023-10-04 06:59:12,162 ERROR [c.c.a.ApiServer] (qtp1239807799-18:ctx-b094b367 ctx-5a43f68f) (logid:a339ed30) unhandled exception executing api command: [Ljava.lang.String;@71c9d354
com.cloud.utils.exception.CloudRuntimeException: Template 205 has not been completely downloaded to zone 1
	at com.cloud.template.TemplateManagerImpl.getTemplateSize(TemplateManagerImpl.java:2020)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:215)
```

In the database, the following entries could be causing the issue

```

MariaDB [cloud]> select * from template_store_ref where template_id=205 \G;
*************************** 1. row ***************************
                  id: 21
            store_id: NULL
         template_id: 205
             created: 2023-10-04 06:30:37
        last_updated: NULL
              job_id: NULL
        download_pct: 0
                size: 0
          store_role: Image
       physical_size: 0
      download_state: BYPASSED
           error_str: NULL
          local_path: NULL
        install_path: NULL
                 url: NULL
               state: Ready
           destroyed: 0
             is_copy: 0
        update_count: 0
             ref_cnt: 0
             updated: NULL
download_url_created: NULL
        download_url: NULL

```

**Expected behavior**

The vm deployment should succeed

**Actual behavior**

The vm deployment is failing

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Fixes #8037

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by shwstppr** [comment]
@blueorangutan package

File: utils/src/main/java/com/cloud/utils/storage/QCOW2Utils.java [hunk 1]
```diff
@@ -114,7 +114,8 @@ private static long getVirtualSizeFromInputStream(InputStream inputStream) throw
     public static long getVirtualSize(String urlStr) {
         try {
             URL url = new URL(urlStr);
-            return getVirtualSizeFromInputStream(url.openStream());
+            boolean isCompressed = !urlStr.endsWith("qcow2");
```
Comment (rohityadavcloud) [inline utils/src/main/java/com/cloud/utils/storage/QCOW2Utils.java:117]: Should we instaed check perhaps for compression extension such as tar.gz, bz2 etc? (what is somebody uploads an .img file which is a qcow2?)

**Comment by DaanHoogland** [comment]
@blueorangutan test rocky8 kvm-rocky8

File: core/src/main/java/org/apache/cloudstack/direct/download/HttpsDirectTemplateDownloader.java [hunk 2]
```diff
@@ -183,7 +186,7 @@ public Long getRemoteFileSize(String url, String format) {
                 SSLContext context = getSSLContext();
                 urlConnection.setSSLSocketFactory(context.getSocketFactory());
                 urlConnection.connect();
-                boolean isCompressed = !url.endsWith("qcow2");
+                boolean isCompressed = UriUtils.COMMPRESSION_FORMATS.stream().anyMatch(f -> !url.toLowerCase().endsWith(f));
                 return QCOW2Utils.getVirtualSize(urlObj.openStream(), isCompressed);
```
Comment (DaanHoogland) [inline core/src/main/java/org/apache/cloudstack/direct/download/HttpsDirectTemplateDownloader.java:190]: should this be a method in `QCOW2Utils`?
the exact same sequence is executed there

**Review by DaanHoogland — APPROVED** [review]
looks good

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Review by rohityadavcloud — APPROVED** [review]

**Pull request merged**